_posts/2018-08-23-make-your-own-cli-with-golang-and-cobra.md

```diff
@@ -0,0 +1,224 @@
+},
+```
+The RunE function also shows the help output if there’s an error.
+This is to show the user how to properly use your command.
+
+At the bottom of the file you’ll see a function called init.
+```go
+func init() {
+   rootCmd.AddCommand(sayCmd)
+
+   // Here you will define your flags and configuration settings.
+
+   // Cobra supports Persistent Flags which will work for this command
+   // and all subcommands, e.g.:
+   // sayCmd.PersistentFlags().String("foo", "", "A help for foo")
+
+   // Cobra supports local flags which will only run when this command
+   // is called directly, e.g.:
+   // sayCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
+}
+```
+Here you can add subcommands and flags.
+By default, the say command is added to the root command, which is exactly what we want.
+Let's add a sub command to say hello.
+```bash
+cobra add sayhello
+```
+Like the say command, sayhello is added to the root command.
+In this case we want it to be added to the say command instead.
+```go
+func init() {
+    sayCmd.AddCommand(sayhelloCmd)
+}
+```
+To make the sayhello print “Hello World”, edit the Run function.
+```go
+Run: func(cmd *cobra.Command, args []string) {
+    fmt.Println("Hello World!")
+},
+```
+Now execute the following inside your projectfolder.
+```bash
+go install
+```
+Followed by
+```bash
+hello-cli say hello
+```
+If everything worked correctly, your CLI should output “Hello World!”.
+
+# Adding flags
+To make your CLI a bit more interesting, we are going to add some flags.
+you can choose between local and persistent ones.
+Local flags are available only for that command, whereas persistent flags are also available for the subcommands of that command.
+For this example we want to greet a person by name.
+We’re going to do this by making a --name flag.
+Navigate to the init function of the sayhello.go file, and add a flag.
+```go
+func init() {
+    rootCmd.AddCommand(sayCmd)
+    sayhelloCmd.Flags().StringP("name","n", "", "Set your name")
+}
+```
+The first string is the full name of the flag, and can be executed with two dashes like “--name”.
+The second string is the short notation, which can be executed with one dash.
+The third one is the default value, and the fourth is a description. 
+
+To make the flag do something we need to add some logic to the Run function.
+```go
+Run: func(cmd *cobra.Command, args []string) {
+    name, _:= cmd.Flags().GetString("name")
+   	if name == "" {
+       		name = "World"
+   	}
+  	fmt.Println("Hallo "+name)
+},
+
+```
+do a quick go install, and check if it works.
+```bash
+hello-cli say hello -n Nick
+```
+The output should be “Hello Nick”. 
+
+# Using environment variables instead of flags
+If you don’t want to pollute your command line, or if you’re working with sensitive data which you don’t want to show up in the history, it’s a good idea to work with environment variables.
+To do this, you can use viper. Viper is another dependency from spf13.
```

Comment:
spf13? The GitHub user https://github.com/spf13?
If so you could link the name to his profile:
```suggestion
To do this, you can use [Viper](https://github.com/spf13/viper){:target="_blank" rel="noopener noreferrer"}. Viper is another dependency from [spf13](https://github.com/spf13){:target="_blank" rel="noopener noreferrer"}.
```